lua/luasnip/util/environ.lua

```diff
@@ -1,190 +1,139 @@
-		util.get_selection()
-end
+local builtin_vars = require("luasnip.util._builtin_vars")
 
-local lazy_vars = {}
+local function tbl_to_lazy_env(tbl)
+	local function wrapper(varname)
+		local val_ = tbl[varname]
+		if type(val_) == "function" then
+			return val_()
+		end
+		return val_
+	end
 
-local Environ = {}
-function Environ:new(pos, o)
-	o = o or {}
-	setmetatable(o, self)
-	_fill_eager_vars(o, pos)
-	return o
+	return wrapper
 end
 
-function Environ:__index(key)
-	local v = lazy_vars[key]()
-	rawset(self, key, v)
-	return v
-end
+local namespaces = {
+	[""] = {
+		init = builtin_vars.eager,
+		vars = tbl_to_lazy_env(builtin_vars.lazy),
+		eager = {},
+	},
+}
 
-function Environ:override(env, new_env)
-	for k, v in pairs(new_env) do
-		env[k] = v
+-- Namespaces allow users to define their own environmet variables
+local function _resolve_namespace_var(full_varname)
+	local parts = vim.split(full_varname, "_")
+	if #parts < 2 then
+		return nil
 	end
-end
+	local nmsp = namespaces[parts[1]]
 
--- Variables defined in https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
-
--- Inherited from TextMate
-function lazy_vars.TM_FILENAME()
-	return vim.fn.expand("%:t")
-end
+	local varname
 
-function lazy_vars.TM_FILENAME_BASE()
-	return vim.fn.expand("%:t:s?\\.[^\\.]\\+$??")
-end
-
-function lazy_vars.TM_DIRECTORY()
-	return vim.fn.expand("%:p:h")
-end
-function lazy_vars.TM_FILEPATH()
-	return vim.fn.expand("%:p")
+	if nmsp then
+		varname = full_varname:sub(#parts[1] + 2)
+	else
+		nmsp = namespaces[""]
+		varname = full_varname
+	end
+	return nmsp.vars(varname)
 end
 
--- Vscode only
+local Environ = {}
 
-function lazy_vars.CLIPBOARD() -- The contents of your clipboard
-	return vim.fn.getreg('"', 1, true)
+-- returns nil, but that should be alright.
+-- If not, use metatable.
+function Environ.is_table(key)
+	return builtin_vars.is_table[key]
 end
 
-local function buf_to_ws_part()
-	local LSP_WORSKPACE_PARTS = "LSP_WORSKPACE_PARTS"
-	local ok, ws_parts = pcall(vim.api.nvim_buf_get_var, 0, LSP_WORSKPACE_PARTS)
-	if not ok then
-		local file_path = vim.fn.expand("%:p")
+function Environ:new(pos, o)
+	o = o or {}
+	setmetatable(o, self)
 
-		for _, ws in pairs(vim.lsp.buf.list_workspace_folders()) do
-			if file_path:find(ws, 1, true) == 1 then
-				ws_parts = { ws, file_path:sub(#ws + 2, -1) }
-				break
+	for ns_name, ns in pairs(namespaces) do
+		local eager_vars = {}
+		if ns.init then
+			eager_vars = ns.init(pos)
+		end
+		for _, eager in ipairs(ns.eager) do
+			if not eager_vars[eager] then
+				eager_vars[eager] = ns.vars(eager)
 			end
 		end
-		-- If it can't be extracted from lsp, then we use the file path
-		if not ok and not ws_parts then
-			ws_parts = { vim.fn.expand("%:p:h"), vim.fn.expand("%:p:t") }
+
+		for name, val in pairs(eager_vars) do
+			if ns_name ~= "" then
+				name = ns_name .. "_" .. name
+			end
+			if val and type(val) ~= "string" and not Environ.is_table(name) then
```

Comment:
Here, `Environ.is_table` should just be the regular `type(val) == "table"`, right? Otherwise this would turn all custom vars which return tables into something like this `"table: 0x7f5a539b5258"` afaict

Reply:
I thought we were using `is_table` to check when expanding, but now that I check is only for the fake expansion, so I will remove the check.

Reply:
Just to expand on that, `fake_expand` (used for docstring, for example) uses a modified `env` which just returns the varname for any variable (`env.TM_SELECTED_TEXT` -> `$TM_SELECTED_TEXT`). Now, for functions to still work correctly we have to return the same type in the fake-expanded snippet.
I don't know how to get that info for custom-env-snippets dynamically, maybe we have to require another `eager`-like-table, which declares which vars are tables

Reply:
Seriously considering requiring that variables are always `string[]` (breaking change, but will make lots of code simpler (no more `if type(var) == "string" then ... else ... end`) and just remove this problem)

Reply:
> If you wanna be able to always show the type, then namespaces will have to be able to provide the type dynamically as well, so it will have to be a "method" on the namespace, and expose the result via `Environ.is_table`

Oh, that's true, but that's really ugly, isn't it?

> f so wdyt if we do it in a different PR (as it does not break the namespaces nor the fake_expand features)

I'd love to, but `fake_expand` will break if 
1. a custom namespace returns a table for some var
2. and that var is used like `env.MY_CUSTOM_TABLE_VAR[1]` in a function/dynamicNode

(since the `__index` from `fake_expand` will return a string, not a table)

What's your opinion on the breaking change?

Reply:
(okay, not `fake_expand` will break, `get_docstring` might)

Reply:
True, I forgot that they can be accessed that way too :thinking: but anyway even with a method or a table for doing it, users could always use the namespace wrong, like the creator of the namespace is thinking a var will be used as a string and the user uses it as a table, or the same namespace creator returning the incorrect type, so I'm not sure what to do about it.

Reply:
What do you think if the `fake_env` vars return always a table with an overridden `__concat` so it behaves as a string? too?

Reply:
What do you think if the fake_env vars return always a table with an overridden __concat so it behaves as a string too?

Reply:
> What do you think if the `fake_env` vars return always a table with an overridden `__concat` so it behaves as a string too?

That would be pretty cool, overriding `__concat` (and `__tostring`) should cover most bases.
But I'm not sure if it's really the best solution, there probably are still exceptions which, however rare, would break (probably creating some confusion along the way) (I'm thinking of stuff like indexing into a table with the value, that would fail no?)

Reply:
Well, yes but then there is no way to fix it besides rendering the environment for real, which is not a good solution either, as namespaces can have secondary effects. So I think is better to make it work for the majority of the users at the cost of facing some issues because of missuses. In any case the worst that can happen is that a `f` or a `d` node felling on the preview, but that's something that can happen with any other misuse isn't it?

Reply:
> Well, yes but then there is no way to fix it besides rendering the environment for real, which is not a good solution either, as namespaces can have secondary effects.

We could require the variables to always be `string[]` (`[str]` :P), that would also prevent this  issue.
But shy of that, yes, overriding `__concat` would work just fine.

>  but that's something that can happen with any other misuse isn't it?

Yeh, also true, but this is something we can prevent, so we should I think.

Reply:
I just added a fake environment that implements `__concat` and `__tostring`

Reply:
Very clean, that should work if we decide to do it that way :+1:
But I'm not sure we should: a clean break, and after it all env-vars are always `string[]`, seems much cleaner than worrying about what is `tostring`d and what isn't.

Also (and I should have realized this sooner), this will only affect the receiving end, we can still allow simple strings as `lazy_vars`, just silently convert it to `string[]` before returning it from `__index` and we should be good to go.

Also (if we do a breaking change), the transition will be very smooth (since it just means some variables need to be indexed with `[1]`), so that should not be a big issue.

Do you see any issues with making the variables uniform? I guess it will make functions that use `env` a _tad_ more verbose, but on the flipside, if `env.SOMEVAR` can be counted on being a table, that could make some stuff easier (if a plugin wants to provide env-variables, to do it properly, there might be a need to list what each var returns (`string` or `string[]`)).
One downside I can see is that there are just very few env-variables which contain multiline text, so the many have to bend to the pressure of the few, which isn't that nice, but I feel like it's the best option we have

Reply:
My concern with that is that it will break all the lua snippet in the while that are already using the vars as strings in any nonsimple way, and given that there are more string vars than lists vars that will bring way more errors than the other approach I assume.

Reply:
Yeah, it might break a fair bit of snippets (I don't think that many, most usage of `env` is probably through `lsp` and `lambda`), but once the bandaid is off luasnip is in a better state than before.
We could also try a deprecation-period with `__concat`, it's a pretty good solution in 99%, but I think tables-only makes more sense in the long run (simpler on our side, no surprises in dynamic/functionNodes when their docstrings are generated).

Reply:
I'm starting to like the metatable more than every-var-is-a-table, it's more user-friendly (well, to the 99%), and the errors caused by it can be silences via `tostring`. We just need a good way to point users unfamiliar with these errors into the right direction (so probably a big warning somewhere in DOC.md for now, logging the errors later).
One more thing: this can't currently handle `env.somevar:match()` or other string-calls (which, I imagine, also make up a big chunk of what's done with the variables). We could override `__index` and return wrapper-functions around the string-methods, which take our fake_env-var and pass its actual real content to the string-method (and returns the string-method's result)